### ISSUE TITLE
Prune mealy machine output

### ISSUE THREAD

**Issue by santolucito** [Barnard-PL-Labs/tsltools#1]
With parsedot (on the parsedot) branch, we can now see the output of the tool in a somewhat human readable form. I'm experiments with @wonhyukchoi and @shmublu we noticed extra transitions. These extras transitions unnecessary and should be pruned. They represent transitions that cannot be taken given the context of the TSL semantics. We potentially need a postprocessing step to prune by looking back to the original TSL formula. There might also be a way to add assumptions to the LTL translation ahead of synthesis to prevent these transitions from being generated in the first place.

Looking at the LTL formula generation code here: https://github.com/Barnard-PL-Labs/tsltools/blob/76ba7c654f665eb78754bd8d8f46d77cd6bb9a28/src/lib/TSL/TLSF.hs#L70

we see that we do not currently generate an ```ASSUME``` block. This might be the way to prune these extras transitions. Since we are changing the approximation from TSL to LTL, we will need to check how this change impacts the formalism presented in the CAV paper (https://www.marksantolucito.com/papers/cav-19.pdf).

The first step is to document what we are seeing on a very small example. @shmublu try to get the tool chain running and recreate the small example we walked through on our call.

**Comment by santolucito** [comment]
Turns out that the newest version of STRIX no longer outputs KISS format files, but instead gives us HOAF (http://adl.github.io/hoaf/). For a first step, we need to get a image file (maybe through a .dot file first) from the .hoaf output.

An example of the new output format is here:

```
REALIZABLE
HOA: v1
tool: "strix" "21.0.0"
States: 2
Start: 0
AP: 6 "p0p0play0note64" "p0p0play0note67" "u0lfo0f1dtoggle0lfo1b" "u0lfo0lfo" "u0fm2synthesis0f1dtoggle0fm2synthesis1b" "u0fm2synthesis0fm2synthesis"
controllable-AP: 2 3 4 5
acc-name: all
Acceptance: 0 t
--BODY--
State: 0 "[0]"
[(!(0 | 1)) & (!(2 | ((4 | !5) | !3)))] 0
[(!(0 | !1)) & (!(2 | ((5 | !4) | !3)))] 0
[(!(1 | !0)) & (!((3 | (4 | !5)) | !2))] 0
[(0 & 1) & (t)] 1
State: 1 "[1]"
[(t) & (t)] 1
--END--
```

@shmublu your first step is to render this as an image using whatever toolchain you can find online

**Comment by santolucito** [comment]
This was generated from the following LTL formula (in .tlsf format):

```
INFO {
  TITLE:       "Converted TSL Specification: test"
  DESCRIPTION: "TSL specification, which has been converted to TLSF."
  SEMANTICS:   Mealy
  TARGET:      Mealy
}
MAIN {
  INPUTS {
    p0p0play0note67;
    p0p0play0note64;
  }

  OUTPUTS {
    u0fm2synthesis0fm2synthesis;
    u0fm2synthesis0f1dtoggle0fm2synthesis1b;
    u0lfo0lfo;
    u0lfo0f1dtoggle0lfo1b;
  }

  ASSUME {
    (G ! ((p0p0play0note67) && (p0p0play0note64)));
  }

  GUARANTEE {
    (G ((! (((u0fm2synthesis0fm2synthesis) && (! (u0fm2synthesis0f1dtoggle0fm2synthesis1b))) <-> ((u0fm2synthesis0f1dtoggle0fm2synthesis1b) && (! (u0fm2synthesis0fm2synthesis))))) && (! (((u0lfo0lfo) && (! (u0lfo0f1dtoggle0lfo1b))) <-> ((u0lfo0f1dtoggle0lfo1b) && (! (u0lfo0lfo))))))) && ((G ((p0p0play0note67) <-> (u0fm2synthesis0f1dtoggle0fm2synthesis1b))) && (G ((p0p0play0note64) <-> (u0lfo0f1dtoggle0lfo1b))));
  }
}
```

**Comment by santolucito** [comment]
As an additional note, this means the parsedot code is now entirely useless and has to be rewritten. awesome...

**Comment by shmublu** [comment]
Yeah, from my limited research there does not appear to be an existing tool that turns it into an image. Whipped up a quick python script, now just need to find out the documentation for how to convert what we have right now into a different format that can be outputted as a PNG of a mealy machine.

**Comment by santolucito** [comment]
I see. Probably the GraphViz library is the best bet. Lmk what you think

**Comment by santolucito** [comment]
I found another option that seems work out of the box. Use the script here: https://github.com/Barnard-PL-Labs/tsltools/blob/parsedot/run_ltlsynt.sh

you will need to install the spot tool (which gives the command line tool ltlsynt). You can install the debian package follwing these instructions: https://spot.lrde.epita.fr/install.html (note that you need to be on ubunutu version >20.04). With this I was able to get the following graph.

![test2](https://user-images.githubusercontent.com/1085500/148593260-77edd2de-735e-4c5e-8247-a884f62b41b3.png)

This was generated from the version of the spec without the assumes added to the tlsf formula. I am not sure what these means in general. Maybe we still should add the assumptions, or maybe ltlsynt is using a different strategy that doesn't generate as many superfluous edges?

**Comment by santolucito** [comment]
So I updated the run_ltlsynt.sh script and the parsedot tool (which is now more like parsehoaf) so that the images are readable again. 
![test2](https://user-images.githubusercontent.com/1085500/148628527-4b5d30b0-b75d-4950-8706-82ff5b44f425.png)

This did create issue #2, but for the purposes of investigating the extra transitions, it should work on small examples at least. Let's see if we can still recreate the "extra transitions" issue as mentioned above, and document here.
